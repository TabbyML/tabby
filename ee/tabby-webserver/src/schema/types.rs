use juniper::{GraphQLInputObject, ID};
use validator::Validate;

#[derive(GraphQLInputObject, Validate)]
pub struct CreateRepositoryProviderInput {
    #[validate(regex(
        code = "displayName",
        path = "crate::schema::constants::REPOSITORY_NAME_REGEX"
    ))]
    pub display_name: String,
    #[validate(length(code = "access_token", min = 10))]
    pub access_token: String,
}

#[derive(GraphQLInputObject, Validate)]
pub struct UpdateRepositoryProviderInput {
    pub id: ID,
    #[validate(regex(
        code = "displayName",
        path = "crate::schema::constants::REPOSITORY_NAME_REGEX"
    ))]
    pub display_name: String,
    #[validate(length(code = "access_token", min = 10))]
    pub access_token: String,
}
