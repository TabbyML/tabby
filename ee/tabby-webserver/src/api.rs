use async_trait::async_trait;
use juniper::{GraphQLEnum, GraphQLObject};
use serde::{Deserialize, Serialize};
use tabby_common::api::{
    code::{CodeSearch, CodeSearchError, SearchResponse},
    event::RawEventLogger,
};
use thiserror::Error;
use tokio_tungstenite::connect_async;

use crate::websocket::WebSocketTransport;

#[derive(GraphQLEnum, Serialize, Deserialize, Clone, Debug)]
pub enum WorkerKind {
    Completion,
    Chat,
}

#[derive(GraphQLObject, Serialize, Deserialize, Clone, Debug)]
pub struct Worker {
    pub kind: WorkerKind,
    pub name: String,
    pub addr: String,
    pub device: String,
    pub arch: String,
    pub cpu_info: String,
    pub cpu_count: i32,
    pub cuda_devices: Vec<String>,
}

#[derive(Serialize, Deserialize, Error, Debug)]
pub enum HubError {
    #[error("Invalid token")]
    InvalidToken(String),

    #[error("Feature requires enterprise license")]
    RequiresEnterpriseLicense,
}

#[tarpc::service]
pub trait Hub {
    async fn register_worker(
        kind: WorkerKind,
        port: i32,
        name: String,
        device: String,
        arch: String,
        cpu_info: String,
        cpu_count: i32,
        cuda_devices: Vec<String>,
        token: String,
    ) -> Result<Worker, HubError>;

    async fn log_event(content: String);

    async fn search(q: String, limit: usize, offset: usize) -> SearchResponse;

    async fn search_in_language(
        language: String,
        tokens: Vec<String>,
        limit: usize,
        offset: usize,
    ) -> SearchResponse;
}

pub fn tracing_context() -> tarpc::context::Context {
    tarpc::context::current()
}

pub async fn create_client(addr: &str) -> HubClient {
    let addr = format!("ws://{}/hub", addr);
    let (socket, _) = connect_async(&addr).await.unwrap();
    HubClient::new(Default::default(), WebSocketTransport::from(socket)).spawn()
}

impl RawEventLogger for HubClient {
    fn log(&self, content: String) {
        let context = tarpc::context::current();
        let client = self.clone();

        tokio::spawn(async move { client.log_event(context, content).await });
    }
}

#[async_trait]
impl CodeSearch for HubClient {
    async fn search(
        &self,
        q: &str,
        limit: usize,
        offset: usize,
    ) -> Result<SearchResponse, CodeSearchError> {
        match self
            .search(tracing_context(), q.to_owned(), limit, offset)
            .await
        {
            Ok(serp) => Ok(serp),
            Err(_) => Err(CodeSearchError::NotReady),
        }
    }

    async fn search_in_language(
        &self,
        language: &str,
        tokens: &[String],
        limit: usize,
        offset: usize,
    ) -> Result<SearchResponse, CodeSearchError> {
        match self
            .search_in_language(
                tracing_context(),
                language.to_owned(),
                tokens.to_owned(),
                limit,
                offset,
            )
            .await
        {
            Ok(serp) => Ok(serp),
            Err(_) => Err(CodeSearchError::NotReady),
        }
    }
}
