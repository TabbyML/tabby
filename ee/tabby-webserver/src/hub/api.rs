use std::net::IpAddr;

use anyhow::Result;
use async_trait::async_trait;
use axum::{headers::Header, http::HeaderName};
use hyper::Request;
use serde::{Deserialize, Serialize};
use tabby_common::{
    api::{
        code::{CodeSearch, CodeSearchError, SearchResponse},
        event::RawEventLogger,
    },
    config::{RepositoryAccess, RepositoryConfig},
};
use tarpc::context::Context;
use tokio_tungstenite::connect_async;

use super::websocket::WebSocketTransport;
use crate::schema::worker::Worker;
pub use crate::schema::worker::WorkerKind;

#[tarpc::service]
pub trait Hub {
    async fn log_event(content: String);

    async fn search(q: String, limit: usize, offset: usize) -> SearchResponse;

    async fn search_in_language(
        language: String,
        tokens: Vec<String>,
        limit: usize,
        offset: usize,
    ) -> SearchResponse;

    async fn list_repositories() -> Vec<RepositoryConfig>;
    async fn create_job_run(name: String) -> i32;
    async fn update_job_output(id: i32, stdout: String, stderr: String);
    async fn complete_job_run(id: i32, exit_code: i32);
}

fn tracing_context() -> tarpc::context::Context {
    tarpc::context::current()
}

fn build_client_request(addr: &str, token: &str, request: ConnectHubRequest) -> Request<()> {
    Request::builder()
        .uri(format!("ws://{}/hub", addr))
        .header("Host", addr)
        .header("Connection", "Upgrade")
        .header("Upgrade", "websocket")
        .header("Sec-WebSocket-Version", "13")
        .header("Sec-WebSocket-Key", "unused")
        .header("Authorization", format!("Bearer {}", token))
        .header("Content-Type", "application/json")
        .header(
            &CLIENT_REQUEST_HEADER,
            serde_json::to_string(&request).unwrap(),
        )
        .body(())
        .unwrap()
}

#[derive(Clone)]
pub struct WorkerClient(HubClient);

pub async fn create_worker_client(
    addr: &str,
    token: &str,
    request: RegisterWorkerRequest,
) -> WorkerClient {
    let request = build_client_request(addr, token, ConnectHubRequest::Worker(request));
    let (socket, _) = connect_async(request).await.unwrap();
    WorkerClient(HubClient::new(Default::default(), WebSocketTransport::from(socket)).spawn())
}

impl RawEventLogger for WorkerClient {
    fn log(&self, content: String) {
        let context = tarpc::context::current();
        let client = self.0.clone();

        tokio::spawn(async move { client.log_event(context, content).await });
    }
}

#[async_trait]
impl CodeSearch for WorkerClient {
    async fn search(
        &self,
        q: &str,
        limit: usize,
        offset: usize,
    ) -> Result<SearchResponse, CodeSearchError> {
        match self
            .0
            .search(tracing_context(), q.to_owned(), limit, offset)
            .await
        {
            Ok(serp) => Ok(serp),
            Err(_) => Err(CodeSearchError::NotReady),
        }
    }

    async fn search_in_language(
        &self,
        language: &str,
        tokens: &[String],
        limit: usize,
        offset: usize,
    ) -> Result<SearchResponse, CodeSearchError> {
        match self
            .0
            .search_in_language(
                tracing_context(),
                language.to_owned(),
                tokens.to_owned(),
                limit,
                offset,
            )
            .await
        {
            Ok(serp) => Ok(serp),
            Err(_) => Err(CodeSearchError::NotReady),
        }
    }
}

#[derive(Serialize, Deserialize)]
pub enum ConnectHubRequest {
    Scheduler,
    Worker(RegisterWorkerRequest),
}

#[derive(Serialize, Deserialize)]
pub struct RegisterWorkerRequest {
    pub kind: WorkerKind,
    pub name: String,
    pub device: String,
    pub arch: String,
    pub cpu_info: String,
    pub cpu_count: i32,
    pub cuda_devices: Vec<String>,
    pub port: u16,
}

impl RegisterWorkerRequest {
    pub fn into_worker(self, addr: IpAddr) -> Worker {
        let port = self.port;
        let addr = format!("http://{addr}:{port}");
        Worker {
            name: self.name,
            kind: self.kind,
            addr,
            device: self.device,
            arch: self.arch,
            cpu_info: self.cpu_info,
            cpu_count: self.cpu_count,
            cuda_devices: self.cuda_devices,
        }
    }
}

pub static CLIENT_REQUEST_HEADER: HeaderName = HeaderName::from_static("x-tabby-client-request");

impl Header for ConnectHubRequest {
    fn name() -> &'static axum::http::HeaderName {
        &CLIENT_REQUEST_HEADER
    }

    fn decode<'i, I>(values: &mut I) -> Result<Self, axum::headers::Error>
    where
        Self: Sized,
        I: Iterator<Item = &'i axum::http::HeaderValue>,
    {
        let mut x: Vec<_> = values
            .map(|x| serde_json::from_slice(x.as_bytes()))
            .collect();
        if let Some(x) = x.pop() {
            x.map_err(|_| axum::headers::Error::invalid())
        } else {
            Err(axum::headers::Error::invalid())
        }
    }

    fn encode<E: Extend<axum::http::HeaderValue>>(&self, _values: &mut E) {
        todo!()
    }
}

#[derive(Clone)]
pub struct SchedulerClient(HubClient);

pub async fn create_scheduler_client(addr: &str, token: &str) -> SchedulerClient {
    let request = build_client_request(addr, token, ConnectHubRequest::Scheduler);
    let (socket, _) = connect_async(request).await.unwrap();
    SchedulerClient(HubClient::new(Default::default(), WebSocketTransport::from(socket)).spawn())
}

#[async_trait]
impl RepositoryAccess for SchedulerClient {
    async fn list_repositories(&self) -> Result<Vec<RepositoryConfig>> {
        Ok(self.0.list_repositories(Context::current()).await?)
    }
    async fn create_job_run(&self, name: String) -> Result<i32> {
        Ok(self.0.create_job_run(tracing_context(), name).await?)
    }
    async fn update_job_output(&self, id: i32, stdout: String, stderr: String) -> Result<()> {
        Ok(self
            .0
            .update_job_output(tracing_context(), id, stdout, stderr)
            .await?)
    }
    async fn complete_job_run(&self, id: i32, exit_code: i32) -> Result<()> {
        Ok(self
            .0
            .complete_job_run(tracing_context(), id, exit_code)
            .await?)
    }
}
