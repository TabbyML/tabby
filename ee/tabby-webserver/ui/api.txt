1:HL["/_next/static/media/86fdec36ddd9097e-s.p.woff2","font",{"crossOrigin":"","type":"font/woff2"}]
2:HL["/_next/static/media/c9a5bc6a7c948fb0-s.p.woff2","font",{"crossOrigin":"","type":"font/woff2"}]
3:HL["/_next/static/css/1a49e85d7522d4c1.css","style"]
0:["na4c0ohcnSYk0g05SAgze",[[["",{"children":["api",{"children":["__PAGE__",{}]}]},"$undefined","$undefined",true],"$L4",[[["$","link","0",{"rel":"stylesheet","href":"/_next/static/css/1a49e85d7522d4c1.css","precedence":"next"}]],"$L5"]]]]
6:I{"id":78495,"chunks":["768:static/chunks/768-5f89fb02d50a441c.js","9109:static/chunks/9109-08bf42cc9883c73f.js","584:static/chunks/584-4a34f53d05daf959.js","3072:static/chunks/3072-c6632acab1bab4ae.js","1424:static/chunks/1424-454dcec055103473.js","2699:static/chunks/2699-d97820a36669949a.js","7070:static/chunks/7070-b53097eca5027517.js","2453:static/chunks/2453-cbd8ebe27ee2c85e.js","7753:static/chunks/7753-850ec6337184216b.js","2445:static/chunks/2445-73fc3774eb68b163.js","3185:static/chunks/app/layout-363768dca7019271.js"],"name":"Providers","async":false}
7:I{"id":81443,"chunks":["2272:static/chunks/webpack-10e7c3c555f15277.js","2971:static/chunks/fd9d1056-1908058eb8be54df.js","7864:static/chunks/7864-52c05c790493c7b0.js"],"name":"","async":false}
8:I{"id":18639,"chunks":["2272:static/chunks/webpack-10e7c3c555f15277.js","2971:static/chunks/fd9d1056-1908058eb8be54df.js","7864:static/chunks/7864-52c05c790493c7b0.js"],"name":"","async":false}
a:I{"id":42761,"chunks":["768:static/chunks/768-5f89fb02d50a441c.js","9109:static/chunks/9109-08bf42cc9883c73f.js","584:static/chunks/584-4a34f53d05daf959.js","3072:static/chunks/3072-c6632acab1bab4ae.js","1424:static/chunks/1424-454dcec055103473.js","2699:static/chunks/2699-d97820a36669949a.js","7070:static/chunks/7070-b53097eca5027517.js","2453:static/chunks/2453-cbd8ebe27ee2c85e.js","7753:static/chunks/7753-850ec6337184216b.js","2445:static/chunks/2445-73fc3774eb68b163.js","3185:static/chunks/app/layout-363768dca7019271.js"],"name":"Toaster","async":false}
5:[["$","meta","0",{"charSet":"utf-8"}],["$","title","1",{"children":"Tabby - API"}],["$","meta","2",{"name":"description","content":"Tabby, an opensource, self-hosted AI coding assistant."}],["$","meta","3",{"name":"theme-color","media":"(prefers-color-scheme: light)","content":"white"}],["$","meta","4",{"name":"theme-color","media":"(prefers-color-scheme: dark)","content":"black"}],["$","meta","5",{"name":"viewport","content":"width=device-width, initial-scale=1"}],["$","meta","6",{"name":"next-size-adjust"}]]
4:[null,["$","html",null,{"lang":"en","suppressHydrationWarning":true,"children":[["$","head",null,{}],["$","body",null,{"className":"font-sans antialiased __variable_aaf875 __variable_2c91d1","children":["$","$L6",null,{"attribute":"class","defaultTheme":"system","enableSystem":true,"children":[["$","div",null,{"className":"flex min-h-screen flex-col","children":["$","$L7",null,{"parallelRouterKey":"children","segmentPath":["children"],"loading":"$undefined","loadingStyles":"$undefined","hasLoading":false,"error":"$undefined","errorStyles":"$undefined","template":["$","$L8",null,{}],"templateStyles":"$undefined","notFound":[["$","title",null,{"children":"404: This page could not be found."}],["$","div",null,{"style":{"fontFamily":"system-ui,\"Segoe UI\",Roboto,Helvetica,Arial,sans-serif,\"Apple Color Emoji\",\"Segoe UI Emoji\"","height":"100vh","textAlign":"center","display":"flex","flexDirection":"column","alignItems":"center","justifyContent":"center"},"children":["$","div",null,{"children":[["$","style",null,{"dangerouslySetInnerHTML":{"__html":"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}"}}],["$","h1",null,{"className":"next-error-h1","style":{"display":"inline-block","margin":"0 20px 0 0","padding":"0 23px 0 0","fontSize":24,"fontWeight":500,"verticalAlign":"top","lineHeight":"49px"},"children":"404"}],["$","div",null,{"style":{"display":"inline-block"},"children":["$","h2",null,{"style":{"fontSize":14,"fontWeight":400,"lineHeight":"49px","margin":0},"children":"This page could not be found."}]}]]}]}]],"notFoundStyles":[],"childProp":{"current":["$","$L7",null,{"parallelRouterKey":"children","segmentPath":["children","api","children"],"loading":"$undefined","loadingStyles":"$undefined","hasLoading":false,"error":"$undefined","errorStyles":"$undefined","template":["$","$L8",null,{}],"templateStyles":"$undefined","notFound":"$undefined","notFoundStyles":"$undefined","childProp":{"current":["$L9",["$","iframe",null,{"className":"grow","src":"/swagger-ui"}],null],"segment":"__PAGE__"},"styles":[]}],"segment":"api"},"styles":[]}]}],["$","$La",null,{"richColors":true,"closeButton":true}],null]}]}]]}],null]
9:null
