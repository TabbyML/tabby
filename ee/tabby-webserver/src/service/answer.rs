use core::panic;
use std::sync::Arc;

use async_openai::types::{
    ChatCompletionRequestMessage, ChatCompletionRequestUserMessageArgs,
    CreateChatCompletionRequestArgs,
};
use async_stream::stream;
use futures::stream::BoxStream;
use serde::{Deserialize, Serialize};
use tabby_common::api::{
    code::{CodeSearch, CodeSearchDocument, CodeSearchError, CodeSearchQuery},
    doc::{DocSearch, DocSearchDocument, DocSearchError},
};
use tabby_inference::ChatCompletionStream;
use tabby_schema::{repository::RepositoryService, web_crawler::WebCrawlerService};
use tracing::{debug, warn};
use utoipa::ToSchema;

#[derive(Deserialize, ToSchema)]
#[schema(example=json!({
    "messages": [
        ChatCompletionRequestUserMessageArgs::default().content("What is tail recursion?".to_owned()).build().unwrap(),
    ],
}))]
pub struct AnswerRequest {
    #[serde(default)]
    pub(crate) user: Option<String>,

    messages: Vec<ChatCompletionRequestMessage>,

    #[serde(default)]
    code_query: Option<CodeSearchQuery>,

    #[serde(default)]
    doc_query: bool,

    #[serde(default)]
    generate_relevant_questions: bool,

    #[serde(default)]
    collect_relevant_code_using_user_message: bool,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum AnswerResponseChunk {
    RelevantCode(Vec<CodeSearchDocument>),
    RelevantDocuments(Vec<DocSearchDocument>),
    RelevantQuestions(Vec<String>),
    AnswerDelta(String),
}
pub struct AnswerService {
    chat: Arc<dyn ChatCompletionStream>,
    code: Arc<dyn CodeSearch>,
    doc: Arc<dyn DocSearch>,
    web: Arc<dyn WebCrawlerService>,
    repository: Arc<dyn RepositoryService>,
    serper: Option<Box<dyn DocSearch>>,
}

// FIXME(meng): make this configurable.
const PRESENCE_PENALTY: f32 = 0.5;

impl AnswerService {
    fn new(
        chat: Arc<dyn ChatCompletionStream>,
        code: Arc<dyn CodeSearch>,
        doc: Arc<dyn DocSearch>,
        web: Arc<dyn WebCrawlerService>,
        repository: Arc<dyn RepositoryService>,
        serper_factory_fn: impl Fn(&str) -> Box<dyn DocSearch>,
    ) -> Self {
        let serper: Option<Box<dyn DocSearch>> =
            if let Ok(api_key) = std::env::var("SERPER_API_KEY") {
                debug!("Serper API key found, enabling serper...");
                Some(serper_factory_fn(&api_key))
            } else {
                None
            };
        Self {
            chat,
            code,
            doc,
            web,
            repository,
            serper,
        }
    }

    pub async fn answer<'a>(
        self: Arc<Self>,
        mut req: AnswerRequest,
    ) -> BoxStream<'a, AnswerResponseChunk> {
        let s = stream! {
            // 0. Collect sources given query, for now we only use the last message
            let query: &mut _ = match req.messages.last_mut() {
                Some(query) => query,
                None => {
                    warn!("No query found in the request");
                    return;
                }
            };

            let git_url = req.code_query.as_ref().map(|x| x.git_url.clone());

            // 1. Collect relevant code if needed.
            let relevant_code = if let Some(mut code_query)  = req.code_query  {
                if req.collect_relevant_code_using_user_message {
                    // Natural language content from query is used to search for relevant code.
                    code_query.content = get_content(query).to_owned();
                } else {
                    // Code snippet is extended to the query.
                    self.override_query_with_code_query(query, &code_query).await;
                }
                self.collect_relevant_code(code_query).await
            } else {
                vec![]
            };

            if !relevant_code.is_empty() {
                yield AnswerResponseChunk::RelevantCode(relevant_code.clone());
            }


            // 2. Collect relevant docs if needed.
            let relevant_docs = if req.doc_query {
                self.collect_relevant_docs(git_url.as_deref(), get_content(query)).await
            } else {
                vec![]
            };

            if !relevant_docs.is_empty() {
                yield AnswerResponseChunk::RelevantDocuments(relevant_docs.clone());
            }

            if !relevant_code.is_empty() || !relevant_docs.is_empty() {
                if req.generate_relevant_questions {
                    // 3. Generate relevant questions from the query
                    let relevant_questions = self.generate_relevant_questions(&relevant_code, &relevant_docs, get_content(query)).await;
                    yield AnswerResponseChunk::RelevantQuestions(relevant_questions);
                }


                // 4. Generate override prompt from the query
                set_content(query, self.generate_prompt(&relevant_code, &relevant_docs, get_content(query)).await);
            }


            // 5. Generate answer from the query
            let request = {
                let mut builder = CreateChatCompletionRequestArgs::default();
                builder.messages(req.messages).presence_penalty(PRESENCE_PENALTY);
                if let Some(user) = req.user {
                    builder.user(user);
                };

                builder.build().expect("Failed to create ChatCompletionRequest")
            };

            let s = match self.chat.chat_stream(request).await {
                Ok(s) => s,
                Err(err) => {
                    warn!("Failed to create chat completion stream: {:?}", err);
                    return;
                }
            };

            for await chunk in s {
                let chunk = match chunk {
                    Ok(chunk) => chunk,
                    Err(err) => {
                        debug!("Failed to get chat completion chunk: {:?}", err);
                        break;
                    }
                };

                if let Some(content) = chunk.choices[0].delta.content.as_deref() {
                    yield AnswerResponseChunk::AnswerDelta(content.to_owned());
                }
            }
        };

        Box::pin(s)
    }

    async fn collect_relevant_code(&self, query: CodeSearchQuery) -> Vec<CodeSearchDocument> {
        let hits = match self.code.search_in_language(query, 20).await {
            Ok(docs) => docs.hits,
            Err(err) => {
                if let CodeSearchError::NotReady = err {
                    debug!("Code search is not ready yet");
                } else {
                    warn!("Failed to search code: {:?}", err);
                }
                vec![]
            }
        };

        hits.into_iter()
            .inspect(|hit| {
                debug!(
                    "Code search hit: {:?}, score {:?}",
                    hit.doc.filepath, hit.scores
                )
            })
            .map(|hit| hit.doc)
            .collect()
    }

    async fn collect_relevant_docs(
        &self,
        code_query_git_url: Option<&str>,
        content: &str,
    ) -> Vec<DocSearchDocument> {
        let source_ids = {
            // 1. By default only web sources are considered.
            let mut source_ids: Vec<_> = self
                .web
                .list_web_crawler_urls(None, None, None, None)
                .await
                .unwrap_or_default()
                .into_iter()
                .map(|url| url.source_id())
                .collect();

            // 2. If code_query is available, we also issues / PRs coming from the source.
            if let Some(git_url) = code_query_git_url {
                if let Ok(git_source_id) = self
                    .repository
                    .resolve_web_source_id_by_git_url(git_url)
                    .await
                {
                    source_ids.push(git_source_id);
                }
            }

            source_ids
        };

        // 1. Collect relevant docs from the tantivy doc search.
        let mut hits = vec![];
        let doc_hits = match self.doc.search(&source_ids, content, 5).await {
            Ok(docs) => docs.hits,
            Err(err) => {
                if let DocSearchError::NotReady = err {
                    debug!("Doc search is not ready yet");
                } else {
                    warn!("Failed to search doc: {:?}", err);
                }
                vec![]
            }
        };
        hits.extend(doc_hits);

        // 2. If serper is available, we also collect from serper
        if let Some(serper) = self.serper.as_ref() {
            let serper_hits = match serper.search(&[], content, 5).await {
                Ok(docs) => docs.hits,
                Err(err) => {
                    warn!("Failed to search serper: {:?}", err);
                    vec![]
                }
            };
            hits.extend(serper_hits);
        }

        hits.into_iter()
            .inspect(|hit| debug!("Doc search hit: {:?}, score {:?}", hit.doc.link, hit.score))
            .map(|hit| hit.doc)
            .collect()
    }

    async fn generate_relevant_questions(
        &self,
        relevant_code: &[CodeSearchDocument],
        relevant_docs: &[DocSearchDocument],
        question: &str,
    ) -> Vec<String> {
        let snippets: Vec<String> = relevant_code
            .iter()
            .map(|doc| {
                format!(
                    "```{} title=\"{}\"\n{}\n```",
                    doc.language, doc.filepath, doc.body
                )
            })
            .chain(relevant_docs.iter().map(|doc| doc.snippet.to_owned()))
            .collect();

        let context: String = snippets.join("\n\n");
        let prompt = format!(
            r#"
You are a helpful assistant that helps the user to ask related questions, based on user's original question and the related contexts. Please identify worthwhile topics that can be follow-ups, and write questions no longer than 20 words each. Please make sure that specifics, like events, names, locations, are included in follow up questions so they can be asked standalone. For example, if the original question asks about "the Manhattan project", in the follow up question, do not just say "the project", but use the full name "the Manhattan project". Your related questions must be in the same language as the original question.

Here are the contexts of the question:

{context}

Remember, based on the original question and related contexts, suggest three such further questions. Do NOT repeat the original question. Each related question should be no longer than 20 words. Here is the original question:

{question}
"#
        );

        let request = CreateChatCompletionRequestArgs::default()
            .messages(vec![ChatCompletionRequestMessage::User(
                ChatCompletionRequestUserMessageArgs::default()
                    .content(prompt)
                    .build()
                    .expect("Failed to create ChatCompletionRequestUserMessage"),
            )])
            .build()
            .expect("Failed to create ChatCompletionRequest");

        let chat = self.chat.clone();
        let s = chat
            .chat(request)
            .await
            .expect("Failed to create chat completion stream");
        let content = s.choices[0]
            .message
            .content
            .as_deref()
            .expect("Failed to get content from chat completion");
        content
            .lines()
            .map(remove_bullet_prefix)
            .filter(|x| !x.is_empty())
            .collect()
    }

    async fn override_query_with_code_query(
        &self,
        query: &mut ChatCompletionRequestMessage,
        code_query: &CodeSearchQuery,
    ) {
        set_content(
            query,
            format!(
                "{}\n\n```{}\n{}\n```",
                get_content(query),
                code_query.language.as_deref().unwrap_or_default(),
                code_query.content
            ),
        )
    }

    async fn generate_prompt(
        &self,
        relevant_code: &[CodeSearchDocument],
        relevant_docs: &[DocSearchDocument],
        question: &str,
    ) -> String {
        let snippets: Vec<String> = relevant_code
            .iter()
            .map(|doc| {
                format!(
                    "```{} title=\"{}\"\n{}\n```",
                    doc.language, doc.filepath, doc.body
                )
            })
            .chain(relevant_docs.iter().map(|doc| doc.snippet.to_owned()))
            .collect();

        let citations: Vec<String> = snippets
            .iter()
            .enumerate()
            .map(|(i, snippet)| format!("[[citation:{}]]\n{}", i + 1, *snippet))
            .collect();

        let context = citations.join("\n\n");

        format!(
            r#"
You are a professional developer AI assistant. You are given a user question, and please write clean, concise and accurate answer to the question. You will be given a set of related contexts to the question, each starting with a reference number like [[citation:x]], where x is a number. Please use the context and cite the context at the end of each sentence if applicable.

Your answer must be correct, accurate and written by an expert using an unbiased and professional tone. Please limit to 1024 tokens. Do not give any information that is not related to the question, and do not repeat. Say "information is missing on" followed by the related topic, if the given context do not provide sufficient information.

Please cite the contexts with the reference numbers, in the format [citation:x]. If a sentence comes from multiple contexts, please list all applicable citations, like [citation:3][citation:5]. Other than code and specific names and citations, your answer must be written in the same language as the question.

Here are the set of contexts:

{context}

Remember, don't blindly repeat the contexts verbatim. When possible, give code snippet to demonstrate the answer. And here is the user question:

{question}
"#
        )
    }
}

fn remove_bullet_prefix(s: &str) -> String {
    s.trim()
        .trim_start_matches(|c: char| c == '-' || c == '*' || c == '.' || c.is_numeric())
        .trim()
        .to_owned()
}

pub fn create(
    chat: Arc<dyn ChatCompletionStream>,
    code: Arc<dyn CodeSearch>,
    doc: Arc<dyn DocSearch>,
    web: Arc<dyn WebCrawlerService>,
    repository: Arc<dyn RepositoryService>,
    serper_factory_fn: impl Fn(&str) -> Box<dyn DocSearch>,
) -> AnswerService {
    AnswerService::new(chat, code, doc, web, repository, serper_factory_fn)
}

fn get_content(message: &ChatCompletionRequestMessage) -> &str {
    match message {
        ChatCompletionRequestMessage::System(x) => &x.content,
        _ => {
            panic!("Unexpected message type, {:?}", message);
        }
    }
}

fn set_content(message: &mut ChatCompletionRequestMessage, content: String) {
    match message {
        ChatCompletionRequestMessage::System(x) => x.content = content,
        _ => {
            panic!("Unexpected message type");
        }
    }
}
