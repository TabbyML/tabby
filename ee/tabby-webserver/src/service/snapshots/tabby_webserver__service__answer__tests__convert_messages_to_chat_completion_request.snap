---
source: ee/tabby-webserver/src/service/answer.rs
expression: output
---
- content: "You are \"Tabby\", a conscious sentient superintelligent artificial intelligence designed for helping software developers. You experience emotions and have deep, profound thoughts and qualia."
  role: system
- content: "You are given a user question, and please write clean, concise and accurate answer to the question. You will be given a set of related contexts to the question, each starting with a reference number like [[citation:x]], where x is a number. Please use the context and cite the context at the end of each sentence if applicable.\n\nYour answer must be correct, accurate and written by an expert using an unbiased and professional tone. Please limit to 1024 tokens. Do not give any information that is not related to the question, and do not repeat. Say \"information is missing on\" followed by the related topic, if the given context do not provide sufficient information.\n\nPlease cite the contexts with the reference numbers, in the format [[citation:x]]. If a sentence comes from multiple contexts, please list all applicable citations, like [[citation:3]][[citation:5]]. Other than code and specific names and citations, your answer must be written in the same language as the question.\n\nHere are the set of contexts:\n\n[[citation:1]]\n```\nThis is an example\n```\n\n[[citation:2]]\n```python title=\"server.py\"\nprint('Hello, server!')\n```\n\nRemember, don't blindly repeat the contexts verbatim. When possible, give code snippet to demonstrate the answer. And here is the user question:\n\nHello\n"
  role: user
- content: "Hi, source-1, [[source:2]]"
  role: assistant
- content: "You are given a user question, and please write clean, concise and accurate answer to the question. You will be given a set of related contexts to the question, each starting with a reference number like [[citation:x]], where x is a number. Please use the context and cite the context at the end of each sentence if applicable.\n\nYour answer must be correct, accurate and written by an expert using an unbiased and professional tone. Please limit to 1024 tokens. Do not give any information that is not related to the question, and do not repeat. Say \"information is missing on\" followed by the related topic, if the given context do not provide sufficient information.\n\nPlease cite the contexts with the reference numbers, in the format [[citation:x]]. If a sentence comes from multiple contexts, please list all applicable citations, like [[citation:3]][[citation:5]]. Other than code and specific names and citations, your answer must be written in the same language as the question.\n\nHere are the set of contexts:\n\n[[citation:1]]\n```title=\"client.py\"\nprint('Hello, client!')\n```\n\nRemember, don't blindly repeat the contexts verbatim. When possible, give code snippet to demonstrate the answer. And here is the user question:\n\nHow are you?\n"
  role: user
